### PR TITLE
[receiver/hostmetrics]  Migrate filesystem scrapper to a new metrics builder

### LINKED ISSUE #7136: Migrate hostmetrics / filesystemscraper to the new Metrics Builder
receiver/hostmetricsreceiver/internal/scraper/filesystemscraper

### PR DESCRIPTION
**Pull request by brushknight**
**Description:** Migrating filesystem scrapper to a new metrics builder without any behavior changes. 
This PR is a copy of [#7173](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/7173) as it was marked as stale.

This PR is needed in order to be able to open [feat: add utilization metric for filesystem](https://github.com/brushknight/opentelemetry-collector-contrib/pull/3) against upstream.

Resolves [#7136](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/7136)

### CODE REVIEW

**Comment by dmitryax** [comment]
Please add a changelog item

File: receiver/hostmetricsreceiver/internal/scraper/filesystemscraper/filesystem_scraper.go [hunk 1]
```diff
@@ -85,8 +99,10 @@ func (s *scraper) Scrape(_ context.Context) (pdata.Metrics, error) {
 
 	if len(usages) > 0 {
 		metrics.EnsureCapacity(metricsLen)
-		initializeFileSystemUsageMetric(metrics.AppendEmpty(), now, usages)
-		appendSystemSpecificMetrics(metrics, now, usages)
+		s.recordFileSystemUsageMetric(now, usages)
+		s.mb.Emit(metrics)
+		s.recordSystemSpecificMetrics(now, usages)
+		s.mb.Emit(metrics)
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/internal/scraper/filesystemscraper/filesystem_scraper.go:105]: s.mb.Emit should be called only once per scrape

File: receiver/hostmetricsreceiver/internal/scraper/filesystemscraper/filesystem_scraper_unix.go [hunk 2]
```diff
@@ -25,22 +25,30 @@ import (
...
-	initializeFileSystemUsageDataPoint(idps.AppendEmpty(), now, deviceUsage.partition, metadata.AttributeState.Used, int64(deviceUsage.usage.Used))
-	initializeFileSystemUsageDataPoint(idps.AppendEmpty(), now, deviceUsage.partition, metadata.AttributeState.Free, int64(deviceUsage.usage.Free))
-	initializeFileSystemUsageDataPoint(idps.AppendEmpty(), now, deviceUsage.partition, metadata.AttributeState.Reserved, int64(deviceUsage.usage.Total-deviceUsage.usage.Used-deviceUsage.usage.Free))
+func (s *scraper) recordFileSystemUsageMetric(now pdata.Timestamp, deviceUsages []*deviceUsage) {
+	for _, deviceUsage := range deviceUsages {
+		s.mb.RecordSystemFilesystemUsageDataPoint(
+			now, int64(deviceUsage.usage.Used), deviceUsage.partition.Device, getMountMode(deviceUsage.partition.Opts), deviceUsage.partition.Mountpoint, deviceUsage.partition.Fstype, metadata.AttributeState.Used)
+		s.mb.RecordSystemFilesystemUsageDataPoint(
+			now, int64(deviceUsage.usage.Free), deviceUsage.partition.Device, getMountMode(deviceUsage.partition.Opts), deviceUsage.partition.Mountpoint, deviceUsage.partition.Fstype, metadata.AttributeState.Free)
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/internal/scraper/filesystemscraper/filesystem_scraper_unix.go:33]: Can you please format these two same way as do for the third one

File: receiver/hostmetricsreceiver/internal/scraper/filesystemscraper/filesystem_scraper.go (hunk 1 above)
Comment (brushknight) [reply receiver/hostmetricsreceiver/internal/scraper/filesystemscraper/filesystem_scraper.go:105]: Then we will change order or metrics, it is acceptable?

File: receiver/hostmetricsreceiver/internal/scraper/filesystemscraper/filesystem_scraper_unix.go (hunk 2 above)
Comment (brushknight) [reply receiver/hostmetricsreceiver/internal/scraper/filesystemscraper/filesystem_scraper_unix.go:33]: sure, will do it on monday

File: receiver/hostmetricsreceiver/internal/scraper/filesystemscraper/filesystem_scraper.go (hunk 1 above)
Comment (dmitryax) [reply receiver/hostmetricsreceiver/internal/scraper/filesystemscraper/filesystem_scraper.go:105]: Yes. Order of metrics is the only change that supposed to happen after switching to the new metrics builder

Comment (brushknight) [reply receiver/hostmetricsreceiver/internal/scraper/filesystemscraper/filesystem_scraper.go]: I added a small helper function for tests `findMetricByName()`, if it is not the right way, let me know and I will think more on a solution.

**Comment by dmitryax** [comment]
Please rebase and run `make generate`. There are some changes to auto-rendered documentation merged recently

File: receiver/hostmetricsreceiver/internal/scraper/filesystemscraper/filesystem_scraper.go (hunk 1 above)
Comment (dmitryax) [reply receiver/hostmetricsreceiver/internal/scraper/filesystemscraper/filesystem_scraper.go]: Sounds good, thanks

**Comment by dmitryax** [comment]
And Changelog still needs to be updated :)

**Review by dmitryax — APPROVED** [review]

**Comment by brushknight** [comment]
Is `load-tests / loadtest` stable? It sometimes fails for me and sometimes not 🤔 I just pushed the same branch and all works - https://github.com/brushknight/opentelemetry-collector-contrib/runs/5213694988?check_suite_focus=true

**Comment by jpkrohling** [comment]
> Is load-tests / loadtest stable? 

They are not always stable.

**Pull request merged**